### ISSUE TITLE
[doc] perldebguts does not have any info how to check that we are in THREADED mode

### ISSUE THREAD

**Issue by KES777** [Perl/perl5#22854]
**Where**
https://perldoc.perl.org/perldebguts

**Description**
The [perldebguts](https://perldoc.perl.org/perldebguts) does not provide any info how to debug threaded perl.

[Here](https://perldoc.perl.org/perldebug) we can find that we can debug a given program using threads: `perl -dt threaded_program_name`

[Here](https://perldoc.perl.org/perl5db.pl#THREADS-SUPPORT) we can find that this support `was built into the debugger as of Perl version 5.8.6`

TODO:
It should be documented how to use that capability.

**Comment by jkeenan** [comment]
> **Where** https://perldoc.perl.org/perldebguts
> 
> **Description** The [perldebguts](https://perldoc.perl.org/perldebguts) does not provide any info how to debug threaded perl.
> 
> [Here](https://perldoc.perl.org/perldebug) we can find that we can debug a given program using threads: `perl -dt threaded_program_name`
> 
> [Here](https://perldoc.perl.org/perl5db.pl#THREADS-SUPPORT) we can find that this support `was built into the debugger as of Perl version 5.8.6`
> 
> TODO: It should be documented how to use that capability.

I myself have never written a threaded Perl program ... and I'll bet that's true for lots of us.  Could you supply a simple threaded program for people to try debugging on?

**Comment by richardleach** [comment]
> It should be documented how to use that capability.

@KES777  - I'm not clear on what you're asking for, in that:

- [perldebug](https://perldoc.perl.org/perldebug) is the user guide for the debugger and is where you found `perl -dt threaded_program_name`.
- [perldebguts](https://perldoc.perl.org/perldebguts) seems to be gory internals details and specific use cases.

Are you looking for:
* _perldebguts_ to include thread-related detail in the "Debugger Internals" section?
*  _perldebguts_ to tell you how to implement threads support when writing your own debugger?
* _perldebug_ to include some of the thread id and breakpoint info mentioned in https://perldoc.perl.org/perl5db.pl#DEBUGGER-INITIALIZATION1 ?
* something else?

If you have a specific use case or problem in mind, knowing what that is might also help us.

**Comment by KES777** [comment]
@jkeenan I was thinking that the script I am debugging is multithreaded. Hopefully it is not. Until I realized that I tried to find what should I implement inside my debugger to support `perl -dt script.pl`. I did not find any mentions on perldebguts page about threads at all.

@richardleach 
> perldebguts to include thread-related detail in the "Debugger Internals" section?

Yes.

> perldebguts to tell you how to implement threads support when writing your own debugger?

Yes.

---

I expect to see on perldebguts page something like this:

> If your script is threaded, then in your debugger you should implement `sub xxx ...` to be able to see `yyy`. To check that your script is threaded please check `yyy`. To check that your debugger was run with `-t` flag please check `$^P` bit to ...

I expect to see cautions about gotchas I should aware of when debugging threaded script. What is the environment differences with regular and threaded debugging.

PS. I am not going to write something threaded or debug something threaded. I just noticed that perldebguts page does not have any info about that and decided to create this ticket.

**Comment by tonycoz** [comment]
perl sets `PERL5DB_THREADED=1` in the environment when invoked with `-dt` and the debugger itself will load threads and threads::shared when that is set, this is described in `perldoc perl5db`.

The debugger detects thread creation in DB::sub by checking if threads::new is being called, so it uses already documented mechanisms.

Settings breakpoints under threads works the same as non-threads (updating `${"_<$filename"}[$line]` (iirc) sets/clears a bit in the dbstate op, yay race conditions.)

The E and e commands, and the thread id in the prompt use the API provided by threads.pm.

Is there other information you're looking for?

This isn't the solution to your issue, but checking exactly what we're missing so we can update it.

**Comment by khwilliamson** [comment]
@KES777  If this information were in perdebguts, would that be what you need?

**Comment by KES777** [comment]
@khwilliamson 
Yes, It would be awesome if `perldebguts` page will have information about `perl debugger+threads`.

@tonycoz Thanks for clarifications. 
> perl sets `PERL5DB_THREADED=1` in the environment when invoked with `-dt` and the debugger itself will load threads and threads::shared when that is set, this is described in `perldoc perl5db`.

`perldoc perlrun` https://perldoc.perl.org/perlrun#-dt does not document `PERL5DB_THREADED=1`. Is this related to perl core or `perl5db.pl` only? If it is done by perlcore then it would be nice to extend `perldoc perlrun` documentation.
I suggest that it would be extremely helpful if perl sets used module for "threading" instead of `1`. Eg. `PERL5DB_THREADED=threads`, `PERL5DB_THREADED=My::Threads::Implementation` etc.

> The debugger detects thread creation in DB::sub by checking if `threads::new` is being called, so it uses already documented mechanisms.

As far, as I know there are two "treads" modules which supports multithreads in perl. Checking `threads::new` in `DB::sub` will detect only one of them. Is my assumption correct?

> Settings breakpoints under threads works the same as non-threads (updating ${"_<$filename"}[$line] (iirc) sets/clears a bit in the dbstate op, yay race conditions.)

I do not think this is a problem. I looks like I can put info about thread from which a break point has been set up into `${"_<$filename"}`. I suppose the main issue would be thread ID detection.

> Is there other information you're looking for?

I believe, the questions that are raised in my head are:
* How to detect that current script is a thread?
* What is its thread ID? And how to detect it?
* How debugger client should distinguish if two threads met the break point?

The last question is not related to `perl` itself, it is more custom debugger implementation, but the first two should be answered somewhere at `perldoc`.

> This isn't the solution to your issue, but checking exactly what we're missing so we can update it.

Yes, Thank you for you clarifications. They gave me some ideas.

PS. I hope my questions/answers are useful as well.

**Comment by tonycoz** [comment]
> > The debugger detects thread creation in DB::sub by checking if `threads::new` is being called, so it uses already documented mechanisms.
> 
> As far, as I know there are two "treads" modules which supports multithreads in perl. Checking `threads::new` in `DB::sub` will detect only one of them. Is my assumption correct?

`perl5db.pl` only supports the threads module.  I don't know of another actively maintained threading module.  threads::lite was last updated in 2016 and is experimental.

The hooking is only done to report the new thread is created, so it isn't critical.

> I believe, the questions that are raised in my head are:
> 
>     * How to detect that current script is a thread?

I assume you mean "is a thread that isn't the main thread".

You can use the threads API, the thread id of the main thread is zero.

```
if (threads->tid == 0) { print "main thread\n" }
```

>     * What is its thread ID? And how to detect it?

`threads->tid` returns the id of the current thread (like `threads->self->tid`)

This is documented in `perldoc threads`.

>     * How debugger client should distinguish if two threads met the break point?

All I can think of it to keep a shared data structure and manage meetings through that, `perl5db.pl` uses `threads::shared` to create such a shared structure, but I don't think it uses it for that.

The threading support of `perl5db.pl` is reasonably well covered by https://perldoc.perl.org/perl5db#THREADS-SUPPORT

**Comment by khwilliamson** [comment]
@KES777  I think this is adequately documented at least somewhere.   I think adding references to those to perldebguts would solve this ticket.  Do you have any suggestions?

**Comment by KES777** [comment]
> Do you have any suggestions?

This one:
>perldoc perlrun https://perldoc.perl.org/perlrun#-dt does not document PERL5DB_THREADED=1. Is this related to perl core or perl5db.pl only? If it is done by perlcore then it would be nice to extend perldoc perlrun documentation.
I suggest that it would be extremely helpful if perl sets used module for "threading" instead of 1. Eg. PERL5DB_THREADED=threads, PERL5DB_THREADED=My::Threads::Implementation etc.
